consensus/reactor.go

```diff
@@ -760,967 +656,688 @@ func (conR *Reactor) gossipVotesForHeight(
+		// peer goroutines and closed r.closeCh, but the router still sends in-flight
+		// peer updates.
+		if !r.IsRunning() {
+			return
+		}
 
-// GetRoundState returns an shallow copy of the PeerRoundState.
-// There's no point in mutating it since it won't change PeerState.
-func (ps *PeerState) GetRoundState() *cstypes.PeerRoundState {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
+		// save the peer if the ID is non-empty and we haven't seen this peer before
+		if psKey != "" && !r.peers.Has(psKey) {
+			peerState := NewPeerState(r.Logger, peerUpdate.PeerID)
+			r.peers.Set(psKey, peerState)
+		}
 
-	prs := ps.PRS // copy
-	return &prs
-}
+		peerState, ok := r.peers.Get(psKey).(*PeerState)
+		if !ok {
+			panic(fmt.Sprintf("peer %v has no state", psKey))
+		}
 
-// ToJSON returns a json of PeerState.
-func (ps *PeerState) ToJSON() ([]byte, error) {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
+		if !peerState.IsRunning() {
+			// Set the peer state's closer to signal to all spawned goroutines to exit
+			// when the peer is removed. We also set the running state to ensure we
+			// do not spawn multiple instances of the same goroutines and finally we
+			// set the waitgroup counter so we know when all goroutines have exited.
+			peerState.broadcastWG.Add(3)
+			peerState.SetRunning(true)
+
+			// start goroutines for this peer
+			go r.gossipDataRoutine(peerState)
+			go r.gossipVotesRoutine(peerState)
+			go r.queryMaj23Routine(peerState)
+
+			// Send our state to the peer. If we're fast-syncing, broadcast a
+			// RoundStepMessage later upon SwitchToConsensus().
+			if !r.WaitSync() {
+				r.sendNewRoundStepMessage(peerState.peerID)
+			}
+		}
 
-	return tmjson.Marshal(ps)
+	case p2p.PeerStatusDown, p2p.PeerStatusRemoved, p2p.PeerStatusBanned:
+		peerState, ok := r.peers.Get(psKey).(*PeerState)
+		if ok && peerState.IsRunning() {
+			// Close the closer to signal to all spawned goroutines to gracefully exit
+			// and we wait for them to all exit before marking the peer state as no
+			// longer running.
+			//
+			// XXX/TODO: It is possible that we can briefly wait here for all goroutines.
+			// Is this problematic?
+			peerState.closer.Close()
+			peerState.broadcastWG.Wait()
+			peerState.SetRunning(false)
+
+			// remove peer from the peer store to prevent unbounded growth
+			r.peers.Delete(psKey)
+		}
+	}
 }
 
-// GetHeight returns an atomic snapshot of the PeerRoundState's height
-// used by the mempool to ensure peers are caught up before broadcasting new txs
-func (ps *PeerState) GetHeight() int64 {
-	ps.mtx.Lock()
-	defer ps.mtx.Unlock()
-	return ps.PRS.Height
-}
+// handleStateMessage handles envelopes sent from peers on the StateChannel.
+// An error is returned if the message is unrecognized or if validation fails.
+func (r *Reactor) handleStateMessage(envelope p2p.Envelope, msgI Message) error {
+	ps, ok := r.peers.Get(string(envelope.From)).(*PeerState)
+	if !ok {
+		panic(fmt.Sprintf("peer %v has no state", envelope.From))
```

Comment:
In this case I think we should just return rather than panicing. If the channel is backed up it's quite possible to receive a message from a peer after the PeerStatusDown message is processed, since these channels are not ordered with respect to each other. This goes for other locations as well.

Reply:
Yeah, good catch.